components/style/properties/shorthand/background.mako.rs

```diff
@@ -12,172 +12,208 @@
-                if let Ok(value) = input.try(|input| background_attachment::parse(context, input)) {
-                    attachment = Some(value);
-                    any = true;
-                    continue
-                }
-            }
-            if origin.is_none() {
-                if let Ok(value) = input.try(|input| background_origin::parse(context, input)) {
-                    origin = Some(value);
-                    any = true;
-                    continue
+        % for name in "image position repeat size attachment origin clip".split():
+            let mut any_${name} = false;
+        % endfor
+
+        let mut background_color = None;
+        let vec = try!(input.parse_comma_separated(|input| {
+            % for name in "image position repeat size attachment origin clip".split():
+                let mut ${name} = None;
+            % endfor
+
+            loop {
+                if background_color.is_none() {
+                    if let Ok(value) = input.try(|input| background_color::parse(context, input)) {
+                        background_color = Some(value);
+                        continue
+                    }
+                } else {
+                    // color can only be the last element
+                    return Err(())
                 }
-            }
-            if clip.is_none() {
-                if let Ok(value) = input.try(|input| background_clip::parse(context, input)) {
-                    clip = Some(value);
-                    any = true;
-                    continue
+                if position.is_none() {
+                    if let Ok(value) = input.try(|input| background_position::single_value::parse(context, input)) {
+                        position = Some(value);
+                        any_position = true;
+
+                        // Parse background size, if applicable.
+                        size = input.try(|input| {
+                            try!(input.expect_delim('/'));
+                            background_size::single_value::parse(context, input)
+                        }).ok();
+                        if let Some(_) = size {
+                            any_size = true;
+                        }
+
+                        continue
+                    }
                 }
+                % for name in "image repeat attachment origin clip".split():
+                    if ${name}.is_none() {
+                        if let Ok(value) = input.try(|input| background_${name}::single_value::parse(context, input)) {
+                            ${name} = Some(value);
+                            any_${name} = true;
+                            continue
+                        }
+                    }
+                % endfor
+                break
             }
-            break
+            Ok((image, position, repeat, size, attachment, origin, clip))
+        }));
+
+        if !(any_image || any_position || any_repeat || any_size ||
```

Comment:
I think this needs to be inside the loop, otherwise you accept things like `background: ,,,red;`